# Building & Deploying a Recommendation System

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("imdb_top_1000.csv")

In [3]:
df.head()

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142 min,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,"28,341,469"
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175 min,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,"134,966,411"
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,"534,858,444"
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202 min,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,"57,300,000"
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96 min,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,"4,360,000"


In [4]:
print("Shape:", df.shape)


Shape: (1000, 16)


In [5]:
print("\nColumns:")
print(df.columns)


Columns:
Index(['Poster_Link', 'Series_Title', 'Released_Year', 'Certificate',
       'Runtime', 'Genre', 'IMDB_Rating', 'Overview', 'Meta_score', 'Director',
       'Star1', 'Star2', 'Star3', 'Star4', 'No_of_Votes', 'Gross'],
      dtype='str')


In [6]:
print("\nInformation:")
print(df.info())


Information:
<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Poster_Link    1000 non-null   str    
 1   Series_Title   1000 non-null   str    
 2   Released_Year  1000 non-null   str    
 3   Certificate    899 non-null    str    
 4   Runtime        1000 non-null   str    
 5   Genre          1000 non-null   str    
 6   IMDB_Rating    1000 non-null   float64
 7   Overview       1000 non-null   str    
 8   Meta_score     843 non-null    float64
 9   Director       1000 non-null   str    
 10  Star1          1000 non-null   str    
 11  Star2          1000 non-null   str    
 12  Star3          1000 non-null   str    
 13  Star4          1000 non-null   str    
 14  No_of_Votes    1000 non-null   int64  
 15  Gross          831 non-null    str    
dtypes: float64(2), int64(1), str(13)
memory usage: 125.1 KB
None


In [7]:
print("\nMissing Values:")
print(df.isnull().sum())


Missing Values:
Poster_Link        0
Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Overview           0
Meta_score       157
Director           0
Star1              0
Star2              0
Star3              0
Star4              0
No_of_Votes        0
Gross            169
dtype: int64


### Observation
#### The dataset contains 1000 movie records with 16 features (columns).
#### The Series_Title column can be used as the unique movie identifier for generating recommendations.
#### The Overview column contains textual descriptions of movies and will be used as the main text feature for the content-based recommendation system.
#### Other useful features such as Genre, Director, IMDB_Rating, and Stars are available and can be used for additional analysis if required.
#### Most columns are complete, but Certificate (101 missing values), Meta_score (157 missing values), and Gross (169 missing values) contain missing data.
#### Since the recommendation system relies on the Overview column, which has no missing values, the missing values in other columns will not affect the recommendation process.
#### The dataset contains a mix of text (string), numeric (float), and integer data types, making it suitable for preprocessing and analysis.

## Task 2

In [8]:
import string
from nltk.corpus import stopwords
import nltk


In [9]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\HP\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [10]:
df['Overview'] = df['Overview'].fillna('')

In [11]:
stop_words = set(stopwords.words('english'))

In [12]:
import re

In [13]:
uppercase_count = df['Overview'].str.contains(r'[A-Z]', regex=True).sum()

print("Rows containing uppercase letters:", uppercase_count)

Rows containing uppercase letters: 1000


In [14]:
punctuation_count = df['Overview'].str.contains(
    f"[{re.escape(string.punctuation)}]",
    regex=True
).sum()

print("Rows containing punctuation marks:", punctuation_count)

Rows containing punctuation marks: 998


In [15]:
print(df['Overview'].head())

0    Two imprisoned men bond over a number of years...
1    An organized crime dynasty's aging patriarch t...
2    When the menace known as the Joker wreaks havo...
3    The early life and career of Vito Corleone in ...
4    A jury holdout attempts to prevent a miscarria...
Name: Overview, dtype: str


In [16]:
def convert_to_lowercase(text):
    return text.lower()

In [17]:
def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))

In [18]:
def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))

In [19]:
def remove_stopwords(text):
    words = text.split()
    words = [word for word in words if word not in stop_words]
    return " ".join(words)

In [20]:
df['clean_text'] = df['Overview'].apply(convert_to_lowercase)

In [21]:
df['clean_text'] = df['clean_text'].apply(remove_punctuation)

In [22]:
df['clean_text'] = df['clean_text'].apply(remove_stopwords)

In [23]:
df[['Overview', 'clean_text']].head()

,Overview,clean_text
0,Two imprisoned men bond over a number of years...,two imprisoned men bond number years finding s...
1,An organized crime dynasty's aging patriarch t...,organized crime dynastys aging patriarch trans...
2,When the menace known as the Joker wreaks havo...,menace known joker wreaks havoc chaos people g...
3,The early life and career of Vito Corleone in ...,early life career vito corleone 1920s new york...
4,A jury holdout attempts to prevent a miscarria...,jury holdout attempts prevent miscarriage just...


## Task 3

In [24]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [25]:
tfidf = TfidfVectorizer(
    max_features=5000,
    ngram_range=(1, 2)
)

In [26]:
tfidf_matrix = tfidf.fit_transform(df['clean_text'])

In [27]:
print("Shape of TF-IDF Matrix:", tfidf_matrix.shape)

Shape of TF-IDF Matrix: (1000, 5000)


In [28]:
print("\nFirst 10 TF-IDF Features:")
print(tfidf.get_feature_names_out()[:10])


First 10 TF-IDF Features:
['00 status' '007' '007 bond' '007 must' '10000' '10000 hed' '100000'
 '100000 deutschmarks' '100th' '100th birthday']


Observation
TF-IDF converted the 1000 cleaned movie overviews into numerical feature vectors.
Each row in the TF-IDF matrix represents a movie, while each column represents a unique word or phrase (feature).
max_features=5000 limits the vocabulary to the 5000 most important features, reducing dimensionality.
ngram_range=(1,2) includes both single words (unigrams) and two-word combinations (bigrams), capturing more context from the movie descriptions.
The resulting TF-IDF matrix will be used in the next task to calculate cosine similarity and generate movie recommendations.

## Task 4

Cosine similarity is used to measure how similar two movie descriptions are based on their TF-IDF vectors. It compares the angle between the vectors instead of their length, making it suitable for text data where documents may have different lengths. A higher cosine similarity score indicates that two movies have more similar content, so it is used to recommend movies with similar overviews.

In [29]:
from sklearn.metrics.pairwise import cosine_similarity

In [30]:
cosine_sim = cosine_similarity(tfidf_matrix)

In [31]:
print("Cosine Similarity Matrix Shape:", cosine_sim.shape)

Cosine Similarity Matrix Shape: (1000, 1000)


In [32]:
print(cosine_sim[0][:10])

[1.         0.         0.         0.         0.         0.03998929
 0.08683547 0.         0.         0.        ]


Observation
The first similarity score is 1.0, which indicates that the movie is 100% similar to itself.
Most of the similarity scores are 0, indicating that those movies have no significant similarity in their descriptions.
A few movies have small similarity values such as 0.0399 and 0.0868, showing that they share some common words or themes with the selected movie.
The similarity score of 0.0868 is higher than 0.0399, meaning that movie is more closely related to the selected movie than the other.
Since the scores are relatively low, the selected movie has limited textual similarity with most other movies in the dataset based on their overviews.
These cosine similarity scores will be used to identify and recommend the most similar movies in the next task.

## Task 5

In [33]:
# Recommendation function
def recommend(item_name, top_n=5):

    if item_name not in df["Series_Title"].values:
        return "Movie not found"

    index = df[df["Series_Title"] == item_name].index[0]

    similarity_scores = list(enumerate(cosine_sim[index]))

    similarity_scores = sorted(similarity_scores, key=lambda x: x[1], reverse=True)

    similarity_scores = similarity_scores[1:top_n+1]

    recommended_movies = []

    for movie in similarity_scores:
        recommended_movies.append(df.iloc[movie[0]]["Series_Title"])

    return recommended_movies

In [36]:
recommend("Titanic")

['The Notebook',
 'Der Himmel über Berlin',
 'Roman Holiday',
 'Laura',
 'Annie Hall']

In [34]:
recommend("The Dark Knight")

['Batman Begins',
 'La battaglia di Algeri',
 'The Dark Knight Rises',
 'Joker',
 'Kill Bill: Vol. 1']

In [35]:
recommend("Avatar")

['Billy Elliot',
 'Ghostbusters',
 'What Ever Happened to Baby Jane?',
 'In Bruges',
 'La dolce vita']

Observation
The recommendation function accepts a movie title as input and returns the top 5 most similar movies.
It uses the cosine similarity matrix to identify movies with the highest similarity scores.
The selected movie is excluded from the recommendations because it always has a similarity score of 1.0 with itself.
The recommended movies are based on similarities in their overview text, making this a content-based recommendation system.
If a movie title is not present in the dataset, the function returns "Movie not found!" to handle invalid input gracefully.

## Task 6